README.rst
Foris controller wireguard module
=================================
This is a wireguard module for foris-controller

Requirements
============

* python3
* foris-controller

Installation
============

	``python3 setup.py install``


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2020-12-11
### Added
- TODO
- TODO


tests/conftest.py
#
# foris-controller-openvpn_client-module
# Copyright (C) 2019 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import pytest
import os

# load common fixtures
from foris_controller_testtools.fixtures import (
    uci_config_default_path,
    cmdline_script_root,
    controller_modules,
    extra_module_paths,
    env_overrides,
    message_bus,
    backend,
)

CMDLINE_SCRIPT_ROOT = os.path.join(os.path.dirname(os.path.realpath(__file__)), "test_root")


@pytest.fixture(scope="module")
def file_root():
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "test_root")


@pytest.fixture(scope="session")
def uci_config_default_path():
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "uci_configs")


@pytest.fixture(scope="session")
def cmdline_script_root():
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), "test_root")


@pytest.fixture(scope="module")
def controller_modules():
    return ["remote", "openvpn_client"]


def pytest_addoption(parser):
    parser.addoption(
        "--backend",
        action="append",
        default=[],
        help=("Set test backend here. available values = (mock, openwrt)"),
    )
    parser.addoption(
        "--message-bus",
        action="append",
        default=[],
        help=("Set test bus here. available values = (unix-socket, ubus, mqtt)"),
    )
    parser.addoption(
        "--debug-output",
        action="store_true",
        default=False,
        help=("Whether show output of foris-controller cmd"),
    )


def pytest_generate_tests(metafunc):
    if "backend" in metafunc.fixturenames:
        backend = metafunc.config.option.backend
        if not backend:
            backend = ["openwrt"]
        metafunc.parametrize("backend_param", backend, scope="module")

    if "message_bus" in metafunc.fixturenames:
        message_bus = metafunc.config.option.message_bus
        if not message_bus:
            message_bus = ["mqtt"]
        metafunc.parametrize("message_bus_param", message_bus, scope="module")


foris_controller_wireguard_module/__init__.py
#
# foris-controller-wireguard-module
# Copyright (C) 2020 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

__version__ = "0.0.0"


foris_controller_modules/wireguard/handlers/openwrt.py
#
# foris-controller-wireguard-module
# Copyright (C) 2020 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import logging
import typing

from foris_controller.handler_base import BaseOpenwrtHandler
from foris_controller.utils import logger_wrapper
from foris_controller_backends.wireguard import WireguardCmds, WireguardFiles, WireguardUci

from .. import Handler

logger = logging.getLogger(__name__)


class OpenwrtWireguardHandler(Handler, BaseOpenwrtHandler):

    asynchronous = WireguardAsync()
    cmds = WireguardCmds()
    files = WireguardFile()
    uci = WireguardUci()

    @logger_wrapper(logger)
    def list(self) -> typing.List[dict]:
        raise NotImplementedError

    @logger_wrapper(logger)
    def set(self, id: str, enabled: bool) -> bool:
        raise NotImplementedError

    @logger_wrapper(logger)
    def add(self, id: str, config: str) -> bool:
        raise NotImplementedError

    @logger_wrapper(logger)
    def delete(self, id: str) -> bool:
        raise NotImplementedError

    @logger_wrapper(logger)
    def generate_ca(self, notify, exit_notify, reset_notify):
        raise NotImplementedError

    @logger_wrapper(logger)
    def get_status(self):
        raise NotImplementedError

    @logger_wrapper(logger)
    def generate_client(self, name, notify, exit_notify, reset_notify):
        raise NotImplementedError

    @logger_wrapper(logger)
    def revoke(self, cert_id):
        raise NotImplementedError

    @logger_wrapper(logger)
    def delete_ca(self):
        raise NotImplementedError

    @logger_wrapper(logger)
    def get_settings(self):
        return self.uci.get_settings()

    @logger_wrapper(logger)
    def server_update_settings(
        self,
        enabled,
        network=None,
        network_netmask=None,
        route_all=None,
        use_dns=None,
        ipv6=None,
        protocol=None,
    ):
        raise NotImplementedError()
        return self.uci.server_update_settings(
            enabled, network, network_netmask, route_all, use_dns, protocol, ipv6
        )

    @logger_wrapper(logger)
    def get_client_config(self, id, hostname=None):
        raise NotImplementedError()
        filtered = [e for e in self.cmds.get_status()["clients"] if e["id"] == id]
        self.uci.update_server_hostname(hostname)  # update hostname in foris uci
        if not filtered:
            return {"status": "not_found"}
        if filtered[0]["status"] == "revoked":
            return {"status": "revoked"}

        options = self.uci.get_options_for_client()
        return {
            "status": "valid",
            "config": self.files.get_config(id=id, hostname=hostname, **options),
            "name": filtered[0]["name"],
        }


foris_controller_modules/wireguard/handlers/mock.py
#
# foris-controller-wireguard-module
# Copyright (C) 2020 CZ.NIC, z.s.p.o. (http://www.nic.cz/)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
#

import logging

from foris_controller.handler_base import BaseMockHandler
from foris_controller.utils import logger_wrapper

from .. import Handler

logger = logging.getLogger(__name__)


class MockWireguardHandler(Handler, BaseMockHandler):
    clients = {}
    ca_generated = False
    clients = []
    current_id = 2
    settings = {
        "enabled": False,
        "ipv6": False,
        "protocol": "udp",
        "network": "10.111.111.0",
        "network_netmask": "255.255.255.0",
        "routes": [{"network": "192.168.1.0", "netmask": "255.255.255.0"}],
        "device": "tun_turris",
        "server_hostname": "",
        "port": 1194,
        "route_all": False,
        "use_dns": False,
    }

    @logger_wrapper(logger)
    def list(self):
        return [
            {"id": k, "enabled": v["enabled"], "running": False}
            for k, v in MockWireguardHandler.clients.items()
        ]

    @logger_wrapper(logger)
    def set(self, id, enabled):

        if id not in MockWireguardHandler.clients:
            return False

        MockWireguardHandler.clients[id]["enabled"] = enabled
        return True

    @logger_wrapper(logger)
    def add(self, id, config):

        if id in MockWireguardHandler.clients:
            return False

        MockWireguardHandler.clients[id] = {"enabled": True, "config": config}
        return True

    @logger_wrapper(logger)
    def delete(self, id):

        if id not in MockWireguardHandler.clients:
            return False

        del MockWireguardHandler.clients[id]
        return True

    @logger_wrapper(logger)
    def generate_ca(self, notify, exit_notify, reset_notify):
        MockOpenvpnHandler.ca_generated = True
        return "%08X" % random.randrange(2 ** 32)

    @logger_wrapper(logger)
    def get_status(self):
        return {
            "status": "ready" if MockOpenvpnHandler.ca_generated else "missing",
            "clients": MockOpenvpnHandler.clients,
        }

    @logger_wrapper(logger)
    def generate_client(self, name, notify, exit_notify, reset_notify):
        MockOpenvpnHandler.clients.append(
            {"id": "%02X" % MockOpenvpnHandler.current_id, "name": name, "status": "valid"}
        )
        MockOpenvpnHandler.current_id += 1

        return "%08X" % random.randrange(2 ** 32)

    @logger_wrapper(logger)
    def revoke(self, cert_id):
        for client in MockOpenvpnHandler.clients:
            if client["id"] == cert_id:
                client["status"] = "revoked"
                return True
        return False

    @logger_wrapper(logger)
    def delete_ca(self):
        MockOpenvpnHandler.ca_generated = False
        MockOpenvpnHandler.clients = []
        MockOpenvpnHandler.current_id = 2
        return True

    @logger_wrapper(logger)
    def get_settings(self):
        return MockOpenvpnHandler.settings

    @logger_wrapper(logger)
    def update_settings(
        self,
        enabled,
        network=None,
        network_netmask=None,
        route_all=None,
        use_dns=None,
        ipv6=None,
        protocol=None,
    ):
        MockOpenvpnHandler.settings["enabled"] = enabled
        if enabled:
            MockOpenvpnHandler.settings["network"] = network
            MockOpenvpnHandler.settings["network_netmask"] = network_netmask
            MockOpenvpnHandler.settings["route_all"] = route_all
            MockOpenvpnHandler.settings["use_dns"] = use_dns
            MockOpenvpnHandler.settings["protocol"] = protocol
            MockOpenvpnHandler.settings["ipv6"] = ipv6

        return True

    @logger_wrapper(logger)
    def get_client_config(self, id, hostname=None):
        filtered = [e for e in MockOpenvpnHandler.clients if e["id"] == id]
        MockOpenvpnHandler.settings["server_hostname"] = hostname if hostname else ""
        if not filtered:
            return {"status": "not_found"}
        if filtered[0]["status"] == "revoked":
            return {"status": "revoked"}
        return {
            "status": "valid",
            "config": "%s" % hostname if hostname else "<guessed_hostname>",
            "name": filtered[0]["name"],
        }
